notes/memory-and-management.md

```diff
@@ -0,0 +1,666 @@
+
+2. The placement of static variables in memory is such that you can support multiple static variables of the same name, so static entities can all have the same name.
+
+3. Since the use of `static` confines `localVar` to each translation unit, each translation unit uses a different `localVar`. Therefore, the output of this program is: 1, 101, 2, 102
+
+However, because of its setup, `static` variables will never be shared by multiple translation units, but as you can see, they bypass the fact that we can't declare multiple `extern` variables with the same name.
+
+You can even place them inside of control structures too. (This implies that you cannot have 2 or more static variables in the same scope). However, this interaction is more about the persistence of these variables. Consider the following modification to `kern.cpp`
+
+```C++
+#include <iostream>
+
+int kernMain() {
+    static int localVar = 1;
+    std::cout << localVar++ << ", ";
+    return 0;
+}
+```
+<div style="text-align: center"> 
+
+Figure 6a - `brian.cpp` 
+
+</div>
+
+What do you think will happen when we run `main()`? Its tempting to think this will function as any other variable now, that we will get the output: 1, 101, 2, 102. However, we know better than that! We know that `static` or `extern` variables are *persistent*. So this variable doesn't get redeclared everytime. What actually happens is that the preprocessor will declare all these variables at compilation. Therefore, whenever we update `localVar` from `kern.cpp`, it really will continue incrementing it by one, and the output is the same as last time: 1, 101, 2, 102. The only difference here is that the scope of `localVar` in `kern.cpp` is limited to `kernMain()` now.
+
+This entire discussion on `extern` v. `static` can be summarized into a table, which also applies to both function and class member definitions.
+
+Table 1: `extern` v. `static`
+Trait|`extern`|`static`
+---|---|---
+Visibility|Global|Local
+Memory Locations|Exactly One|At Least One
+Containable?|No|Yes
+
+## The Heap
+
+All of us have memories we cherish. Whether it is something like playing with friends or getting a cool present, most of these memories get stored in the hippocampus, which is in the cerebrum. In C/C++, that would be the heap. It is a unique area in the memory where the coder can choose to send memory here, and is exclusively for variables.
+
+We also have memories we need to remember, such as knowledge for an upcoming test. To cultivate such memories, we need to do things like study. The heap also comes with similar responsibilities. Just as you are the librarian of your brain, your code must keep the heap in good shape and up to date.
+
+Finally, there are some memories we wish to forget. For example, when you accidently scream in front of your entire class. In C/C++, you can actually do this with heap memory, and it is as vital that you do this when needed.
```

Comment:
sus

It would also be nice to quickly go over why we would want the heap, for example:
- We need t to store values that persist across multiple different scopes/functions, etc.
- We need to allocate a huge block of memory (since heap is much larger than the stack)